hibernate-core/src/main/java/org/hibernate/query/sqm/tree/jpa/CriteriaParameterCollectQueryWalker.java

```diff
@@ -93,4 +97,40 @@ public Object visitNamedParameterExpression(SqmNamedParameter expression) {
 		return param;
 	}
 
+	@Override
+	public Object visitFunction(SqmFunction sqmFunction) {
+		for ( Object argument : sqmFunction.getArguments() ) {
+			if ( argument instanceof SqmFunction) {
+				visitFunction( (SqmFunction) argument );
+			}
+			else if ( argument instanceof JpaCriteriaParameter ) {
+				visitJpaCriteriaParameter( (JpaCriteriaParameter<?>) argument );
+			}
+			else if ( argument instanceof SqmPositionalParameter ) {
+				visitPositionalParameterExpression( (SqmPositionalParameter) argument );
+			}
+			else if ( argument instanceof SqmNamedParameter ) {
+				visitNamedParameterExpression( ( SqmNamedParameter ) argument );
+			}
+		}
+		return sqmFunction;
+	}
+
+	@Override
+	public Object visitSearchedCaseExpression(SqmCaseSearched<?> expression) {
+		expression.getWhenFragments().forEach( whenFragment -> {
+			whenFragment.getPredicate().accept( this );
+			whenFragment.getResult().accept( this );
+		} );
+		expression.getOtherwise().accept( this );
+		return expression;
+	}
+
+	@Override
+	public Object visitBinaryArithmeticExpression(SqmBinaryArithmetic expression) {
+		expression.getLeftHandOperand().accept( this );
+		expression.getRightHandOperand().accept( this );
+		return expression;
+	}
```

Comment:
I was torn apart by the decision which method to put into base parent class and which ones to be overridden. For instance, it seems our current base or default implementation aims to explore thoroughly, so we should visit as much as possible there? But we can easily think of child class whose requirement conflicts with this, for instance:

- some logging class that needs to input tab or end-of-line before visiting another method;
- some comment specific class that needs to only focus on comment node and ignore all the others and wanna be performant.

The downside to put such stuff into the base parent class is the logic coupling due to inheritance (Effective Java has a good chapter on the danger of it) and it would be so difficult to change the parent class without ripple effect on all the other child classes which relies on it.
Some random thoughts. I am gonna change as you suggested in this PR.

Reply:
I move methods to BaseSemanticQueryWalker as you suggested.